Enhancements to prediction module for assimilation DPLE's

bradyrx opened the pull request:
* Add simple functions to `prediction.py` to facilitate statistics on assimilation DPLE's. 

* Update DPLE notebook with a lot more context, different tests, etc.

aaronspring commented:
NEED TO EITHER WORK OUT THEORY OR FUNCTIONS HERE -- SOMETHING IS WRONG

What is wrong?
- do you mean the ACC values that I got vs the quick solution you wrote?
- can you maybe modifiy your example and use xskillscore's pearson_r (its very fast and will then also work on mapped data)

CESM LE will be good for bootstrapping. Thanks for adding.

Minor comments:
- Ensemble Mean Lead Year Predictions plot would be nice with legend for LYs

aaronspring commented:
ah, ok, you just refered to the persistence function, right? I'm a unsure how persistence works. What I implemented so far (but not nice and not for maps) is damped persistence, when an anomaly just falls back to the mean with lag1-corr damping.

Do you know some nice book or lecture notes about persistence and forecasting in general? So far, I just ask people and google, but never had a lecture on this

bradyrx commented:
This is a work in progress, so this was just a note to remind me where to pick up today, don't look too much into it yet! I am including basic metrics that we tend to use with the assimilation-perspective (persistence, skill, predictability, etc.) and wanted to get practice with `proplot` which I am really starting to like.

> What is wrong?

My quick ACF function I wrote doesn't match what an out-of-the-box ACF suggests (and I believe the one I loaded from `statsmodels`). This doesn't mean yet that our autocorr functions are wrong, I just wrote this right before bed.

> can you maybe modifiy your example and use xskillscore's pearson_r (its very fast and will then also work on mapped data)

Yes, I can do this. In the long run though, my issue with the `xskillscore` pearson_r is that it doesn't account for autocorrelation. My function `xr_eff_pearsonr` in the stats module computes an "effective"  sample size which modifies the p-value. This is important to think about for statistical significance, especially with short time series like we have in the prediction runs. 

Although this is something I will implement once I get moving with my research since it will be more important to be careful about there.

(See the methods section of my paper: https://www.biogeosciences-discuss.net/bg-2018-415/)

> I'm a unsure how persistence works. What I implemented so far (but not nice and not for maps) is damped persistence, when an anomaly just falls back to the mean with lag1-corr damping.

I think it is similar to this. We use persistence as a reference point of prediction skill for the assimilation perspective. Basically you just lag-1, lag-2, etc. correlate the hindcast (assimilation) to itself. The question is: "Is my initialized DPLE better at prediction than just expecting next year to be like the last?"

> Do you know some nice book or lecture notes about persistence and forecasting in general? So far, I just ask people and google, but never had a lecture on this

Unfortuantely not. I have been doing the same thing as you. Some DOI's of papers I have referenced:

* Kirtman, Ben, et al. "Near-term climate change: projections and predictability." (2013).

* 10.5194/esd-2018-73

* 10.1175/BAMS-D-12-00241.1

* 10.1175/2009BAMS2778.1

* 10.1175/BAMS-D-17-0098.1

bradyrx commented:
I also need to catch up to you on what all these different metrics mean. I imagine I have only been using `ACC`, as that's all I have seen being used around me. I.e., I generate anomalies of the DPLE (initialized), CESM-LE (uninitialized), and hindcast (assimilation). I then correlate them at different lags and so on to come up with persistence predictability, uninitialized predictability, and initialized predictability. I can then also correlate DPLE anomalies to true observations to get "skill."

I am unsure where to use all of these other metrics you implemented. Any references on those? I think you might have some in the code documentation...

aaronspring commented:
> Yes, I can do this. In the long run though, my issue with the xskillscore pearson_r is that it doesn't account for autocorrelation. My function xr_eff_pearsonr in the stats module computes an "effective" sample size which modifies the p-value. This is important to think about for statistical significance, especially with short time series like we have in the prediction runs.

pearson_r just calculates the correlation according to the formula. I dont understand yet fully whats the effect of the autocorrelation to the pearson correlation. As far as I can see it in your xr_eff_pearson, you weight the results, but all you use is the first output from xr_eff_pearson. 
To me its not too important to use xskillscore, but from my experience its super fast and fast than the scipy stuff when I used it on large maps (1000 years, 256x220 grid).
I'm just not sure whether for the skill scores we need this effective pearson_r. I couldnt find a link to that in the book I use if I use one:
> Forecast Verification
A Practitioner’s Guide in Atmospheric Science
SECOND EDITION
Edited by
Ian T. Jolliffe
David B. Stephenson

So I dont see you thats needed for the persistence forecast. But I guess we are approach persistence skill calculation from different angles.

bradyrx commented:
> pearson_r just calculates the correlation according to the formula. I dont understand yet fully whats the effect of the autocorrelation to the pearson correlation. As far as I can see it in your xr_eff_pearson, you weight the results, but all you use is the first output from xr_eff_pearson.

Correct, I'll edit my code to use the xskillscore pearson_r to make it quicker. The effective pearson_r doesn't change the r-value (correlation coefficient) but changes the value of r that is needed to be significant (i.e., it changes the p-value).

Basically, when you smooth time series or use ocean/climate data, there's generally a lot of autocorrelation. This artificially lowers the p-value, since it seems like there's a lot of highly correlated values. However, points [A, B, C, D, E] might be highly correlated and thus not independent from one another.

What the effective sample size does is it only calculates the significance with statistically independent samples from the time series (maybe now it's only [A, C, E] or [A, E]). Hope this makes sense! It's to be more rigorous with statistical significance in your correlations. Again, this won't affect the correlation coefficient, but could bring your p-value above 0.05 if that's the value you're using.

bradyrx commented:
Thanks for all these references, I'll look into them.

aaronspring commented:
Usually I use bootstrapping and don’t rely on any assumptions of the form of the distribution. 

This auto correlation is very important to keep in mind I am convinced. The hindcast bootstrapping function doesn’t do that yet. It just shuffles the time series whereas it should some just take other evolutions of that variable which will have autocorrelation as it’s ocean data. 

I am in discussion with Leo about that.

bradyrx commented:
f87dc414823c74792e9aadd52513b4c0eb22a138 creates an explicit `xr_cross_corr` for cross-correlating two time series. It uses the `pearson_r` from `xskillscore` since that's much faster. 

I moved the effective p-value to a separate underscore function that is only called in the cross_corr. I also updated your `xr_corr` to `xr_autocorr` and removed mine. In yours, I give an option to return the p-value and call the xskillscore p_value function for that. I'm not sure yet if we should do effective sample size here.

The effective samples formula used in my paper/Nikki's paper is derived from cross-correlating. I'm not sure that it makes sense to use it in `xr_autocorr` since we're literally computing autocorrelation there.. if that makes sense. I'll think about it for the future.

`SkillMetrics` looks very nice and is good to keep in mind.

bradyrx commented:
@aaronspring, I agree on both points. 1c1a943 changes the function to `xr_corr`, but just keep in mind the distinction between this and `xr_autocorr`. It now accepts two DataArrays instead of one dataset.

bradyrx commented:
Yeah I used that to derive the p-value in `xr_autocorr`. I didn't use it in `xr_corr`, because it is only a two-tailed version. I wanted to provide the option for one-tailed and also use the effective sample size.

aaronspring commented:
Do you want to keep this pull request open or should we merge?
What else would you like to add before?

bradyrx commented:
I am working on adding a persistence function, since that's very useful for CESM-DPLE. 

This is very simple if you have one timeseries (e.g. global average):
```
from statsmodels.tsa.stattools import acf as acf
# compute acf from lead year 1 to 10 (this is persistence skill)
acf(hind, nlags=11)[1::] 
```

However, statsmodels has a hard lock on anything above 1D. This makes it very hard to run the acf on a grid of data. Just trying to write my own code to allow an acf on a grid of data...

I've stripped the statsmodels functions down to exactly what I need, but I'm working on making `np.correlate` work across a grid of data.

bradyrx commented:
@aaronspring , I'm struggling with `xr.apply_ufunc` like usual...

see if you have any thoughts here on how not to make it error:
```
from numpy import correlate as npc
da = xr.DataArray(np.random.rand(10,5), dims=['time','x'])
xr.apply_ufunc(npc, da, da, input_core_dims=[['time'],['time']],
               output_core_dims=[[]], kwargs={'mode': 'full'})
```

I just want that last function to return something of size (19, 5). A single time series (e.g. da.isel(x=0)) returns something of length 19.

aaronspring commented:
How about 
xr.concat([xr_autocorr(ds,lag=i) for i in range(10)],’year’)

I know it’s powerful but I also regularly fail with ufunc

bradyrx commented:
I can't call `xr_autocorr` for this one. Persistence is based off the ACF (autocorrelation function) which takes autocovariance into account. It's subtle but a very important difference.

aaronspring commented:
Ah ok. 

Another way could be to create a persistence dataset (equal values for ym data) and then just compute skill for that. 
That’s how I approached persistence forecast and it’s skill so far. That doesn’t care for autocorrelation and the results as seen in my pm notebook don’t look so nice. I would expect something more smooth...

bradyrx commented:
Unfortunately I don't think that's a great approach for DPLE with a hindcast/reconstruction. The ACF approach is critical, but folks make it hard to adapt to 2+ dimensions. I implemented a 1D version now and will finish the DPLE notebook before merging.

I am going to post on stackexchange for some help with that numpy.correlate! Spent too much time on it.

aaronspring commented:
Ok. Just saw the persistence forecast function. I would rather call it persistence forecast skill function as it give the skill of the persistence forecast and not the forecast itself. 
Your persistence forecast function should start with 1 for lag 0 and then decrease?

I searched the web and I couldn’t find any explanation for persistence forecast except for everything stays as it is. I will keep asking in-house 

Terminolog: 
- forecast: one member, has units of here C

bradyrx commented:
Correct, I'll change to persistence forecast skill. Correct it should start with 1 for lag 0 then decrease, but we only care about lag 1 on so I drop that from the output.

aaronspring commented:
http://www.cpc.ncep.noaa.gov/products/people/wd51hd/book/review/chapter8.pdf

Some explanation. Persistence forecast approaches ACF

aaronspring commented:
Can you post the stackexchange link once you ,post it please?

bradyrx commented:
@aaronspring  -- here's the stackoverflow link: https://stackoverflow.com/questions/54120129/compute-acf-for-a-grid-of-time-series-in-python?noredirect=1#comment95080938_54120129

This might help you see very clearly what I'm trying to speed up.

bradyrx commented:
@aaronspring, here is the biggest issue. Again, basic persistence should just be autocorrelation at various lags for a timeseries. If I run our `xr_autocorr` function, I get the blue line. Using `statsmodels` ACF function with biased and unbiased turned on, I get the black and red lines. I'm not sure what's causing the subtle differences.

This isn't an issue with our `autocorr` function, as any corr function from python creates the same line. Something is happening under the hood in ACF. This means that the vectorization/speed issue shouldn't be a problem, because we can just use `autocorr`, but I need to figure out why these are different.

![](https://i.imgur.com/HZLJjn5.png)

bradyrx commented:
Okay, it looks like by default ACF (and autocorr in MATLAB) uses an FFT method to make processing faster. However, it looks like you sacrifice some accuracy, especially when you have short time series like these. I am going to update the persistence function to use `xr_autocorr` which will make things much faster anyways.

bradyrx commented:
Okay, I fleshed out the DPLE notebook with a ton more information and using my new functions. I copied your original notebook because I still want to go through it to better understand the metrics and maybe rewrite some functions based on that! 

An initial look shows a very similar result between my `DP_compute_skill` (which is just various lag correlations) and your `hindcast_compute`. Although there were subtle differences, despite us both using pearson_r, so I need to look into what's causing that.

bradyrx commented:
Sorry for all the comments; working a lot on this today. So I reviewed the first bit of your code and I see what's causing the discrepancy. 


A stripped down version of your `hindcast_compute` assuming we choose _pearson_r and _e2a:
```
def hindcast_compute(DPE, assim, metric=_pearson_r, comparison=_e2a, time_dim='ensemble'):
    if metric.__name__ in ['_pearson_r', '_rmse']:
        fct, truth = comparison(DPE, assim)
        *es = metric(fct, truth,'ensemble')
        return res
```

Currently, this correlates lead year 1, 2, 3, 4, ... to the hindcast time series without shifting it at all. Thus, it isn't really predicting anything. What I have currently implemented is the proper way to do this and uses `pearson_r` from `xskillscore`, which I agree is a good package:

```
res = xr.concat([xr_corr(DPE.isel(time=i), assim, lag=i, dim='ensemble') for i in range(0, nlags)], 'lag')
```

I'll update `prediction.py` using these ideas (a `hindcast_compute` like function). But feel free to review any code you'd like if you have the time.

aaronspring commented:
I just spoke to my friend about persistence forecast:
- I approached persistence forecast from a totally different perspective: I first looked how a persistence forecast timeseries would look like and then compute the skill of that.
- What I know understood: persistence forecast is basically just shifting the reference
- concerning the magic of statsmodels ACF: here we usually just use the ACC on those shifted timeseries
- Now I get your approach!

aaronspring commented on esmtools/prediction.py:
```diff
@@ -609,9 +604,78 @@ def PM_compute(ds, control, metric=_pearson_r, comparison=_m2m, anomaly=False,
 # PERSISTANCE FORECASTS
 #--------------------------------------------#
+def persistence_forecast_skill(reconstruction, nlags):
+    """
```
it would be nice to have this generalized for more metrics.
I have a function in mind where we can choose the metric of the persistence forecast. See my comment in https://github.com/bradyrx/esmtools/pull/21
Maybe we can adapt `autocorr` in a way that it can be uses as `_pearson_r`. Generally, I think we should use the metrics consistently, meaning using the same math for persistence forecast skill and DPLE/PM-based skill.

bradyrx replied:
This is now generalized in 0f0191e. The user can input a control run, obs, or reconstruction and select pearson_r or rmse.

aaronspring commented on esmtools/prediction.py:
```diff
@@ -609,9 +604,78 @@ def PM_compute(ds, control, metric=_pearson_r, comparison=_m2m, anomaly=False,
+        members.
+    reference : xarray object
+        Reconstruction or observations over same time period
+    nlags : int (default length of `time` dim)
```
really like the idea of saving computation time here, when nlags=2

bradyrx replied:
I think this isn't relevant now that I have made changes to have a compute function for reconstruction systems.

aaronspring commented on esmtools/prediction.py:
```diff
@@ -609,9 +604,78 @@ def PM_compute(ds, control, metric=_pearson_r, comparison=_m2m, anomaly=False,
+    """
+    _check_xarray(dp)
+    _check_xarray(reference)
+    if 'member' in _get_dims(dp):
```
in my setup I use the _m2e in this case. So this function is a ensemble mean against assimilation reference skill.

aaronspring commented on esmtools/prediction.py:
```diff
@@ -609,9 +604,78 @@ def PM_compute(ds, control, metric=_pearson_r, comparison=_m2m, anomaly=False,
+    if nlags is None:
+        nlags = dp['time'].size
+    skill = xr.concat([xr_corr(dp.isel(time=i), reference, lag=i,
+                dim='ensemble') for i in range(0, nlags)], 'lag')
```
It depends on what we would like to see on the x-axis of plots. 
I would give it the name 'time' as in the lead time before, but its a matter of taste.

bradyrx replied:
I switched to 'lead time'

aaronspring commented on esmtools/prediction.py:
```diff
@@ -609,9 +604,78 @@ def PM_compute(ds, control, metric=_pearson_r, comparison=_m2m, anomaly=False,
+    N = dp['ensemble'].size
+    if nlags is None:
+        nlags = dp['time'].size
+    skill = xr.concat([xr_corr(dp.isel(time=i), reference, lag=i,
```
it would be nice to have this generalized for more metrics, at least also _rmse which follows the same requirements as _pearson_r.

I have a function in mind where we can choose the metric for skill forecast. See my comment in #21 

Some thing like. and then we can replace _rmse or xr_autocorr with metrics.

```skill=xr.concat([_rmse(dp.isel(time=i),reference,'ensemble') for i in range(DPE.time.size)],'time')```
But in this line I forgot to shift the `ensemble` dimension of assim (I think thats needed for _pearson_r and _rmse to work. That feature is built-in `xr_autocorr` and `xr_corr` and I tried to get it out into `_shift`)
I think if you can somehow rewrite xr_autocorr in the same way as xskillscore.pearson_r or xs.rmse, then we are there.

aaronspring replied:
something like: this now even works for both comparison types

```python
def hindcast_compute(ds, reference, metric=_pearson_r):
    skilllist=[]
    for i in range(ds.time.size):
        a,b=_shift(ds.isel(time=i),assim,lag=i,dim='ensemble')
        if 'member' in ds:
            a,b=xr.broadcast(a,b)
            a=a.stack(svd=('ensemble','member'))
            b=b.stack(svd=('ensemble','member'))
            skilllist.append(metric(a,b,'svd'))
        else:
            skilllist.append(metric(a,b,'ensemble'))
    skill = xr.concat(skilllist,dim='time')
    return skill

hindcast_compute(DPE, assim)['SST'].plot()
hindcast_compute(DPE.mean('member'), assim)['SST'].plot()
```

Another future step (next PR) would be to take only a few selected ensembles (init years) and see if anomalous starting conditions influences the skill.

bradyrx replied:
Thanks for all these comments. I am going to go through them and try to incorporate what I can. At some point I will probably merge what is working since this pull request is getting huge. Then we can open a new one that focuses on the things I missed.

aaronspring replied:
That’s the m2a example here in if member in ds and the ensemble mean version without computing it below. We are very close to finishing this. The I will just go over the PM stuff later. But great work so far.

aaronspring replied:
I think with this we can merge then.

bradyrx commented:
Okay, I am going to just comment my thoughts with each big change I make since there's a lot going on here. 0f0191e creates a generalized persistence function called `compute_persistence()`. 

* I removed the `ds` input you had to avoid confusing the user. One should only need to put in the time series that they are computing persistence from (hindcast, control run, observations). I forced the input of nlags, which isn't too cumbersome for someone to see how long their DPLE run is and then input that.

* Added a `_get_metric_function()` which converts an input metric string (e.g. 'pearson_r') to the actual function name ('_pearson_r'). This makes sure the user doesn't have to import underscore functions to input into our functions, since by definition we don't want that of underscore functions. This adds a little clunkiness where we have to call eval(metric)(input, input, input), but I think it's fine (see `compute_persistence` for an example).

* Changed `year` output of persistence to `lead year` for clarity. I think this is better than year and lag.

Other thoughts:

1. Your _shift method is the perfect solution to our issues.

2. I like looping through the lags. It's a lot more readable than list comprehension.

bradyrx commented:
Now I updated the `PM_compute` function.

* I changed the name to `compute_perfect_model`. I would like all of our high-level functions to start with 'compute_'. I would also like to be as explicit as possible for the user.

* I created a similar `_get_comparison_function()` and updated `_get_metric_function()` to include all possible metrics. Now you can input strings into `compute_perfect_model` for the metric and comparison so the user doesn't have to import the underscore functions.

aaronspring commented on esmtools/prediction.py:
```diff
@@ -735,12 +810,16 @@ def compute_persistence(reference, nlags, metric='pearson_r'):
     _check_xarray(reference)
     metric = _get_metric_function(metric)
+    if metric not in [_pearson_r, _rmse]:
+        raise ValueError("""Please select between the following metrics:
```
In the future we probably just try to get a MSE function like in xskillscore. And then I can get rid of my complicated other version.

bradyrx replied:
Agreed. I just have persistence and reconstruction compute working for pearson_r and rmse. We can worry about other metrics later and bring in the other packages you've listed.

aaronspring commented on esmtools/prediction.py:
```diff
@@ -638,21 +707,24 @@ def PM_compute(ds, control, metric=_pearson_r, comparison=_m2m, anomaly=False,
 
 
-def DP_compute_skill(dp, reference, nlags=None):
+def compute_reconstruction(ds, reference, metric='pearson_r', nlags=None):
```
Why call it reconstruction? 
We could add here the the comparison later. I’m not sure you will want to analyse also the skill of single starting years (still unsure how to do that) or the skill of ensemble members instead of just the ensemble mean forecast. 

If people want to have an ensemble mean forecast the could also just enter the mean into the function. Or there should be at least a flag to computer single member forecast and mean over the member dim.

aaronspring replied:
Maybe compute reference? Or just compute skill? Or just compute?

bradyrx replied:
> Why call it reconstruction?

I'll switch to compute_reference, but in the future we should have a PR that carefully looks at docstring language.

To me, CESM-DPLE is a hindcast-based prediction system. A hindcast being a model simulation of the past. However, this language is troublesome because the atmosphere community uses hindcast to mean actual predictions of the past.

So the language is better as "reconstruction" for a historical simulation. The CESM-DPLE reconstruction is forced by CORE (heat, momentum, freshwater flux) but has NO data assimilation. So this is not an assimilation either.

Make sense? I agree we go with "reference" because users can put in observations here to compute true skill against observations.

bradyrx commented:
Lastly, ed61ff2  creates a function `compute_reconstruction` which works like `compute_perfect_model`. 

This introduces two new comparisons (adapted from your original notebook): 

* e2r : ensemble mean to reference

* m2r : each ensemble member to the reference

It also allows for `pearson_r` and `rmse` from `xskillscore` to be selected.

For now, this locks the user into choosing pearson r or rmse, but in future PR's we can allow for other metrics for a reconstruction-based system. I'm happy with just pearson r and rmse for now.

bradyrx commented:
@aaronspring, I am running the example notebooks and then will merge.

aaronspring commented:
I don’t have the overview completely but it seems very good. I think you are good to merge once you put in every thing for now. The rest will be a new PR

aaronspring commented:
Can only I merge this? If not please go ahead and merge once your are done. Thanks for pushing this.
